File: clase-03-22/tads/openHashMap.cpp
#pragma once

#include "map.cpp"
#include <assert.h>

// esto es para poder usar hash y el equals
// seguramente tambien lo quieran usar para el hash cerrado
#include <functional>

#include <vector>

template <class K, class V, class Hash = std::hash<K>>
class OpenHashMap : public Map<K, V>
{
public:
  OpenHashMap(int expectedSize)
  {
    this->arrSize = expectedSize * 2 - 1;
    this->arr = new std::vector<Pair<K, V>> *[this->arrSize];
    this->elemCount = 0;
  }

  void insert(K key, V value) override
  {
    if (this->contains(key))
    {
      this->delete (key);
    }
    this->elemCount++;
    int pos = this->findPos(key);
    std::vector<Pair<K, V>> *list = this->arr[pos];
    if (vector == nullptr)
    {
      list = new std::vector<Pair<K, V>>();
      arr[pos] = list;
    }
    Pair p = new Pair(key, value);
    list->insert(p);
  }

  bool contains(K key) override
  {
    int pos = this->findPos(key);
    auto list = this->arr[pos];
    if (list == null)
    {
      return false;
    }
    Pair p = new Pair(key);
    // aca llamarian a su lista a preguntarle si tiene al par p
    for (Pair<K, V> other : list)
    {
      if (other == p)
      {
        return true;
      }
    }
    return false;
  }

  V get(K key) override
  {
    assert(this->contains(key));
    return V();
  }

  void delete (K key) override
  {
  }

  int size() override
  {
    return this.elemCount;
  }

  bool isEmpty() override
  {
    return this.elemCount == 0;
  }

private:
  Hash hashCode;
  // es una array de punteros a vectoras
  std::vector<Pair<K, V>> **arr;
  int arrSize;
  int elemCount;

  int findPos(K key)
  {
    int h = hashCode(key);
    int pos = h % this->arr.length;
    return pos;
  }
};

template <class K, class V, class Equals = std::equal_to<K>>
struct Pair
{
  K key;
  V value;

  Pair(K key, V value)
  {
    this->key = key;
    this->value = value;
  };

  Pair(K key)
  {
    this->key = key;
  }

  bool operator==(Pair<K, V> other)
  {
    return equals(this->key, other->key);
  }

private:
  Equals equals;
};

File: clase-03-22/tads/OpenHashMap.java
package tads;

import java.util.ArrayList;

public class OpenHashMap<K, V> implements Map<K, V> {
    // ArrayList<Pair>[] arr
    private Object[] arr;
    private int elemCount;

    private class Pair {
        public K key;
        public V value;

        Pair(K key, V value) {
            this.key = key;
            this.value = value;
        }

        public Pair(K key) {
            this.key = key;
        }

        @Override
        public boolean equals(Object other) {
            return this.key.equals(((Pair) other).key);
        }

        // c++
        /*
         * bool operator==(Pair<K,V> other)
         * {
         * this->key = other.key;
         * }
         */
    }

    @Override
    public void insert(K key, V value) {
        this.elemCount++;
        int pos = this.findPos(key);
        var list = (ArrayList<Pair>) this.arr[pos];
        if (list == null) {
            list = new ArrayList<Pair>();
            arr[pos] = list;
        }
        Pair p = new Pair(key, value);
        if (list.contains(p)) {
            this.elemCount--;
            list.remove(p);
        }
        list.add(p);
    }

    @Override
    public boolean contains(K key) {
        int pos = this.findPos(key);
        var list = (ArrayList<Pair>) this.arr[pos];
        if (list == null) {
            return false;
        }
        Pair p = new Pair(key);
        return list.contains(p);
    }

    @Override
    public V get(K key) throws Exception {
        // TODO Auto-generated method stub
        return null;
        /*
         * if(!list.contains(p)){
         * throw new Exception(
         * "la tabla no contiene al emeneto" + key.toString());
         * }
         */
    }

    @Override
    public void delete(K key) {
        // TODO Auto-generated method stub

    }

    @Override
    public int size() {
        return this.elemCount;
    }

    @Override
    public boolean isEmpty() {
        return this.elemCount == 0;
    }

    public OpenHashMap(int expectedSize) {
        this.arr = new Object[expectedSize * 2 - 1];
        this.elemCount = 0;
    }

    private int findPos(K key) {
        int h = key.hashCode();
        int pos = h % this.arr.length;
        return pos;
    }

}
